components/calendar/src/chinese.rs

```diff
@@ -0,0 +1,1044 @@
+        let day = last_day - Chinese::chinese_new_moon_before(last_day.as_moment()) + 1;
+        (month, day as u8)
+    }
+}
+
+impl Calendar for Chinese {
+    type DateInner = ChineseDateInner;
+
+    // Construct a date from era/month codes and fields
+    fn date_from_codes(
+        &self,
+        _era: types::Era,
+        year: i32,
+        month_code: types::MonthCode,
+        day: u8,
+    ) -> Result<Self::DateInner, CalendarError> {
+        ArithmeticDate::new_from_lunar_codes(self, year, month_code, day).map(ChineseDateInner)
+    }
+
+    // Construct the date from an ISO date
+    fn date_from_iso(&self, iso: Date<Iso>) -> Self::DateInner {
+        let fixed = Iso::fixed_from_iso(iso.inner);
+        Chinese::chinese_date_from_fixed(fixed).inner
+    }
+
+    // Obtain an ISO date from a Chinese date
+    fn date_to_iso(&self, date: &Self::DateInner) -> Date<Iso> {
+        let fixed = Chinese::fixed_from_chinese_date_inner(*date);
+        Iso::iso_from_fixed(fixed)
+    }
+
+    //Count the number of months in a given year, specified by providing a date
+    // from that year
+    fn days_in_year(&self, date: &Self::DateInner) -> u32 {
+        Self::days_in_provided_year(date.0.year)
+    }
+
+    fn days_in_month(&self, date: &Self::DateInner) -> u8 {
+        Self::month_days(date.0.year, date.0.month)
+    }
+
+    #[doc(hidden)] // unstable
+    fn offset_date(&self, date: &mut Self::DateInner, offset: DateDuration<Self>) {
+        date.0.offset_date(offset)
+    }
+
+    #[doc(hidden)] // unstable
+    #[allow(clippy::field_reassign_with_default)]
+    /// Calculate `date2 - date` as a duration
+    ///
+    /// `calendar2` is the calendar object associated with `date2`. In case the specific calendar objects
+    /// differ on date, the date for the first calendar is used, and `date2` may be converted if necessary.
+    fn until(
+        &self,
+        date1: &Self::DateInner,
+        date2: &Self::DateInner,
+        _calendar2: &Self,
+        _largest_unit: DateDurationUnit,
+        _smallest_unit: DateDurationUnit,
+    ) -> DateDuration<Self> {
+        date1.0.until(date2.0, _largest_unit, _smallest_unit)
+    }
+
+    /// Obtain a name for the calendar for debug printing
+    fn debug_name(&self) -> &'static str {
+        "Chinese"
+    }
+
+    /// The calendar-specific year represented by `date`
+    fn year(&self, date: &Self::DateInner) -> types::FormattableYear {
+        Self::format_chinese_year(date.0.year)
+    }
+
+    /// The calendar-specific month code represented by `date`
+    /// since the Chinese calendar has leap months, it is necessary to track
+    /// whether a month is a leap month, as well as whether it comes after a
+    /// leap month in the current calendar year. To accomplish this, the Chinese
+    /// calendar month codes appends "L" (for "Leap") to a month code if it is a leap month,
+    /// and appends "I" (for "Increment") to a month code if it occurs after a leap month.
```

Comment:
issue: I don't think this `I` scheme is part of the month code design, with month codes you have `M01` and `M01L` and that's it. The number is the 1-indexed ordinal number of that month in a common (i.e. non-leap) year, and L means it occurs after said month. (For leap months at the beginning of the year, the code is M00L.)

The draft Temporal specification, which was designed in concert with ICU4X's month code design, is [here](https://tc39.es/proposal-temporal/#sec-temporal-calendardatemonthcode).

Calculating the month code from the ordinal month, and vice versa, _must_ involve calculation of the leap month shape of that particular year.

(Which in the future we can optimize by caching)

Reply:
I just made up the `I` pattern since I didn't know there was any official version. I've reworked it to be compliant with the Temporal specification you linked; converting between ordinals and month codes now requires calculating the leap month for a given year, but that should be fine for now, and like you said, it can be improved later.